File: Task/Meissel-Mertens-constant/C++/meissel-mertens-constant.cpp
#include <cmath>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <vector>

std::vector<double> list_prime_reciprocals(const int32_t& limit) {
	const int32_t half_limit = ( limit % 2 == 0 ) ? limit / 2 : 1 + limit / 2;
	std::vector<bool> composite(half_limit);
	for ( int32_t i = 1, p = 3; i < half_limit; p += 2, ++i ) {
		if ( ! composite[i] ) {
			for ( int32_t a = i + p; a < half_limit; a = a + p ) {
				composite[a] = true;
			}
		}
	}

	std::vector<double> result(composite.size());
	result[0] = 0.5;
	for ( int32_t i = 1, p = 3; i < half_limit; p += 2, ++i ) {
		if ( ! composite[i] ) {
			result.emplace_back(1.0 / p);
		}
	}
	return result;
}

int main() {
	std::vector<double> prime_reciprocals = list_prime_reciprocals(100000000);
	const double euler = 0.577215664901532861;
	double sum = 0.0;
	for ( double reciprocal : prime_reciprocals ) {
		sum += reciprocal + log(1.0 - reciprocal);
	}

	const double meissel_mertens = euler + sum;
	std::cout << "The Meissel-Mertens constant = " << std::setprecision(8) << meissel_mertens << std::endl;
}

File: Task/Meissel-Mertens-constant/Rust/meissel-mertens-constant.rs
fn list_prime_reciprocals(limit: i32) -> Vec<f64> {
    let half_limit = if limit % 2 == 0 { limit / 2 } else { 1 + limit / 2 };
    let mut composite = vec![false; half_limit as usize];

    for i in 1..half_limit as usize {
        let p = 3 + 2 * (i as i32 - 1);
        if !composite[i] {
            let mut a = i + p as usize;
            while a < half_limit as usize {
                composite[a] = true;
                a += p as usize;
            }
        }
    }

    let mut result = Vec::with_capacity(half_limit as usize);
    result.push(0.5); // Add 1/2 for prime number 2

    for i in 1..half_limit as usize {
        let p = 3 + 2 * (i as i32 - 1);
        if !composite[i] {
            result.push(1.0 / p as f64);
        }
    }

    result
}

fn main() {
    let prime_reciprocals = list_prime_reciprocals(100000000);
    let euler = 0.577215664901532861;

    let sum: f64 = prime_reciprocals.iter()
        .map(|&reciprocal| reciprocal + (1.0 - reciprocal).ln())
        .sum();

    let meissel_mertens = euler + sum;
    println!("The Meissel-Mertens constant = {:.8}", meissel_mertens);
}
